Support ChaChaPoly1305 on Android if available

vcsjones opened the pull request:
This implements ChaChaPoly1305 on Android, if it is available.

Contributes to #52482.

/cc @jkoritzinsky @steveisok

vcsjones committed:
Initial support for ChaChaPoly1305 on Android

vcsjones committed:
Implement IsSupported correctly for ChaCha20Poly1305 on Android

vcsjones commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_cipher.c:
```diff
@@ -138,7 +157,15 @@ ARGS_NON_NULL_ALL static int32_t ReinitializeCipher(CipherCtx* ctx)
+        // ChaChaPoly1305 does not allow configuration of the tag length and
+        // we don't permit user-defined counters, so we can use IvParameterSpec
+        // for ChaChaPoly1305.
+        if (strcmp("ChaCha20/Poly1305/NoPadding", ctx->type->name) == 0)
```
This feels a little gross to me, happy to change this if someone has a better idea.

ChaChaPoly1305 does have an authentication tag, but that algorithm forbids a tag length other than 16 bytes. Because of that, ChaChaPoly1305 on Android can use the `IvParameterSpec`.

vcsjones replied:
I did a bit of refactoring here and I'm happier with where it ended up.

jkoritzinsky commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_cipher.c:
```diff
@@ -75,6 +76,24 @@ ARGS_NON_NULL_ALL static jobject GetAlgorithmName(JNIEnv* env, CipherInfo* type)
+
+    // If we were able to call Cipher.getInstance without an exception, like NoSuchAlgorithmException,
+    // then the algorithm is supported.
+    return CheckJNIExceptions(env) ? FAIL : SUCCESS;
```
You should use `TryClearJNIExceptions` here since an exception is an expected scenario.

vcsjones replied:
Ah. That does the same think as `Check` just without the `ExceptionDescribe`. Thanks!

jkoritzinsky commented on src/libraries/System.Security.Cryptography.Algorithms/tests/ChaCha20Poly1305Tests.cs:
```diff
@@ -447,6 +447,10 @@ public static void CheckIsSupported()
             }
+            else if (PlatformDetection.IsAndroid)
+            {
+                expectedIsSupported = true;
```
We should have this check if it's supported at runtime so the unit tests will still pass when run on older supported simulator versions.

vcsjones replied:
I didn't actually see a good way to check that `IsSupported` is implemented sanely on Android. I don't see an API anywhere that tells you the API Level (sdk version) of the current version of Android.

It doesn't make sense to check this the same way that `IsSupported` is implemented (seeing if we can instantiate a ChaCha20/Poly1305/NoPadding cipher), since we'd essentially be doing `Assert.Equal(ChaChaPoly1305.IsSupported, ChaChaPoly1305.IsSupported)`.

`OperatingSystem.IsAndroidVersionAtLeast` doesn't help us here, either. It depends on the Linux kernel version, not the OS version (and sometimes the API Level bumps for the same OS via OTA updates). 

I didn't see `android.os.Build.VERSION.SDK_INT` exposed in a manged API, either. So: I ended up shelling out to `getprop`, for now.

Again, definitely open to suggestions for improvements or ideas.

vcsjones committed:
Don't describe the exception in an expected-exception path.

vcsjones committed:
Detect Android SDK version for ChaChaPoly1305 IsSupported.

vcsjones committed:
Refactor "HasTag" implementation.

The CIPHER_HAS_TAG was used to determine if the GCMParameterSpec
configuration is needed for variable length tags.

While ChaCha20Poly1305 has authentication tags, it does not permit a tag
length other than 16 bytes, so there is nothing to configure.

This renames the CIPHER_HAS_TAG to be more specific that the cipher
supports more than one tag length, and removes it from ChaCha20Poly1305.

This simplifies the IV initialization a bit.

GrabYourPitchforks commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_cipher.c:
```diff
@@ -75,6 +76,24 @@ ARGS_NON_NULL_ALL static jobject GetAlgorithmName(JNIEnv* env, CipherInfo* type)
+    if (!algName)
+        return FAIL;
+
+    jobject cipher = (*env)->CallStaticObjectMethod(env, g_cipherClass, g_cipherGetInstanceMethod, algName);
```
Why does this line appear different than - say - line 106?

vcsjones replied:
To `GRef` creates a global reference, meaning you want it to live until you call `ReleaseGRef`. Line 106 creates a cipher instance and then stashes it in the `CipherCtx`. From the [JNI docs](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#referencing_java_objects):

> The JNI divides object references used by the native code into two categories: local and global references. Local references are valid for the duration of a native method call, and are automatically freed after the native method returns. Global references remain valid until they are explicitly freed.

But in this particular case, we don't need the `Cipher` instance to live, we just care if it throws. Since we don't need the object instance, it doesn't need to get promoted to a global reference, it can stay as a local and we can remove the local reference.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.Algorithms/tests/ChaCha20Poly1305Tests.cs:
```diff
@@ -456,5 +462,22 @@ public static void CheckIsSupported()
             Assert.Equal(expectedIsSupported, ChaCha20Poly1305.IsSupported);
         }
+
+        private static int GetAndroidSdkVersion()
```
Does it make sense to move this to a shared helper library? If you don't think anybody else has an immediate need for it, disregard this comment.

vcsjones replied:
I'm not sure. I think the majority of tests / implementation should do feature detection (like how `IsSupported` is actually implemented). However we want to test the `IsSupported` implementation against some other mechanism for correctness, and in Android that would be the API version.

I hesitated to put this somewhere easily accessible off of `PlatformDetection` or similar, otherwise it might be mistaken as a good way to guard tests / functionality when we only need it here for a very specific purpose.

vcsjones replied:
Oh, you already fixed it. Thank you!

GrabYourPitchforks reviewed (commented):
I'm not an Android interop expert, but the control flow LGTM.

jkoritzinsky commented on src/libraries/System.Security.Cryptography.Algorithms/tests/ChaCha20Poly1305Tests.cs:
```diff
@@ -447,6 +448,11 @@ public static void CheckIsSupported()
+            else if (PlatformDetection.IsAndroid)
+            {
+                // Android with API Level 28 is the minimum API Level support for ChaChaPoly1305.
+                expectedIsSupported = GetAndroidSdkVersion() >= 28;
```
I'd rather we import the Android interop files to do the `Interop.Crypto.CipherIsSupported(Interop.Crypto.EvpChaCha20Poly1305());` check instead of the Android API version. Importing the interop APIs is the same pattern used with AES, RC2, etc.

vcsjones replied:
Hm. Maybe that kind of defeats the purpose of this test.

The tests are already conditional on ChaChaPoly1305.IsSupported which relies on `CipherIsSupported`. The purpose of this test is to test the `IsSupported` property against some other measure of "is this cipher supported"?

vcsjones replied:
That said, I am happy to make the change if we are sure testing the `IsSupported` property against the API Level is not desirable.

jkoritzinsky replied:
I missed that this was a test of the IsSupported property itself. This is a good way of doing it then I think.

vcsjones commented:
Browser / wasm failure is most likely unrelated.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ChaCha20Poly1305.Android.cs:
```diff
@@ -0,0 +1,159 @@
+                Interop.Crypto.CipherUpdateAAD(_ctxHandle, associatedData);
+            }
+
+            if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, plaintext, out int plaintextBytesWritten, ciphertext))
+            {
+                CryptographicOperations.ZeroMemory(plaintext);
+                throw new CryptographicException();
+            }
+
+            if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, plaintext.Slice(plaintextBytesWritten), out int bytesWritten, tag))
+            {
+                CryptographicOperations.ZeroMemory(plaintext);
+                throw new CryptographicException();
+            }
+
+            plaintextBytesWritten += bytesWritten;
+
+            if (!Interop.Crypto.EvpCipherFinalEx(
+                _ctxHandle,
+                plaintext.Slice(plaintextBytesWritten),
+                out bytesWritten))
+            {
+                CryptographicOperations.ZeroMemory(plaintext);
+                throw new CryptographicException(SR.Cryptography_AuthTagMismatch);
+            }
```
This order of operations makes no sense to me.  But the tests say it works... so.. OK?

vcsjones replied:
This is mostly modeled off of how the GCM implementation does it, but this is my understanding for decrypt

There are two calls to `EvpCipherUpdate` because Android's APIs accept the authentication tag appended to the ciphertext  (`ciphertext || tag`). Android doesn't know how long your plaintext is, so the first update doesn't "decrypt" the number of bytes it was passed, it needs to wait until `final` is called to say, "okay, I've sent all of the ciphertext including the tag".

So the first call sends all of the ciphertext. It may not write the last 16 bytes because it doesn't know that it has reached the tag.

Next, we write the tag. The can also write to the plaintext buffer since the API figured out, "oh wait, you sent me more data, so that data I was hanging on to incase it was the tag can't be the tag, so write it".

Then we call final and that wraps everything up.

bartonjs replied:
Oh, right, "input" is last.  "Here's the AD", "here's the ciphertext", "here's the tag", "I'm done".  Seems obvious, just that it's out/in not in/out.

bartonjs reviewed (commented):
Looks reasonable to me (modulo one nod and smile).  I'm happy so long as @jkoritzinsky is happy with the JNI code.
